### DISCUSSION TITLE
Update SQLAlchemy session objects otherwise than by executing a session-enabled query

### DISCUSSION THREAD

**Discussion by karolinepauls** [sqlalchemy/sqlalchemy#9289] [Usage Questions]
I have a NIH pseudo-ORM to work with which is able to map a row into a class. SQLAlchemy is already used for connection management but nothing more. The codebase has no notion object identity or session.  I want to be able to use SQLAlchemy declarative and its sessions.

I was able to automatically generate a usable SQLAlchemy Declarative schema. I'm using SQLAlchemy 1.4 with legacy cursors. Migration to 2.0 is unrealistic, as it would require rewriting thousands of usage sites of legacy cursors.

The problem is that using Declarative along with the current/legacy codebase would cause SQLAlchemy session desync. So far, I found that I can load resulting rows from raw SQL into SQAlchemy session objects like this: 

```
 >>> user = db.session.query(Users).from_statement(text("UPDATE users SET lastname = 'WillBeChanged' WHERE email = 'my@email.com' RETURNING *")).populate_existing().one()
 ```
 
This however won't be sufficient, since the current/legacy codebase executes queries "by hand" in thousands of places. 
 
I wanted to exploit the fact that the current/legacy codebase tends to issue `f"SELECT {self.columns}"` and `f"RETURNING {self.columns}"` in queries, which return rows later fed into a single base `from_db_row` method mapping a single DB row into a specific model instance. If I were able to achieve the same behaviour as `populate_existing`, but in my `from_db_row` method, I would be able to make current/legacy code update SQLAlchemy session state whenever it accesses the database. That would solve all of session desync problems of INSERT, UPSERT, and UPDATE (not DELETE).

_Currently, I am not trying to map existing classes with SQLAlchemy. All I want is to update objects in the current session as the legacy code updates them via raw SQL. I would prefer SQLAlchemy objects to be separate from the legacy NIH ORM, since I want to enforce strong conventions in SQLAlchemy code._

Thanks for your input.

PS: hack idea: Since the current code already subclasses the SQLAlchemy `Connection`, I could hack it to execute something like

(pseudocode, I don't remember the API exactly)
```
class Connection(sqlalchemy.core.Connection):
    def execute(self, clause, params):
        if not ('SELECT' in clause or 'RETURNING' in clause):
            return super().execute(clause, params)

        sqlalchemy_model = self.figure_it_out_somehow()
        sqlalchemy_result = db.session.query(sqlalchemy_model).from_statement(clause, params).populate_existing().one()
        reconstructed_rows = SomeFakeCursor(r.somehow_get_tuple() for r in sqlalchemy_result)
        return reconstructed_rows
```

**Comment by zzzeek** [comment]
hi -

your post uses terminology that I'm not familiar with, including "legacy cursors" and "session desync".

My guess is that by "legacy cursors" you actually mean [legacy Query](https://docs.sqlalchemy.org/en/20/orm/queryguide/query.html#legacy-query-api) , referring to the ORM Query object.     The phrase "legacy" explicitly means "is not going away" and 2.0 has not changed anything at all about `Query` vs. 1.4, you can still call upon `session.query().populate_existing()` in 2.0 and no such code has to change.

Additionally, the "populate existing" feature remains a first class feature as well, to use it with newer style queries you use the [populate_existing execution option](https://docs.sqlalchemy.org/en/20/orm/queryguide/api.html#populate-existing).

**Comment by karolinepauls** [comment]
Sorry for confusion, by "session desync" I mean raw SQL changing rows in the database which don't get reflected in the identity map, which can lead to all sorts of problems.

My problem is that the only convenient place to update the objects in the current SQLAlchemy session is not the place where queries are executed.

The current "pattern" in our "DIY ORM" is: `connection.execute(text(...))` -----rows-----> `map(from_db_row, ...)` ----> DIYORMModels.

`connection.execute(text(...))` is done in a thousands of places and isn't realistically change them all, however `from_db_row` is only one place in code and it knows which of the current "DIY ORM" models it is mapping rows to. It is then trivial to know which SQLAlchemy models it corresponds to.

So I would essentially like to be able to tell SQLAlchemy to update its objects in the session without marking them as "dirty".

**Reply by zzzeek** [reply]
OK was not clear if this is a migration from 1.4 to 2.0 issue, or a new feature in your program issue.   As yet I dont see any limitation with using 2.0 and dont know what "legacy cursors" refers towards.

as far as the interplay between these ``connection.execute()`` calls and your `Session`, things depend highly on specifics.  The usual way to have what we might refer towards as "out of band" SQL DML ( insert/update/delete) to be reflected in objects that are in a `Session` is to simply expire them, using `Session.expire_all()` or `Session.expire(object)`.    There's also the usage of populate_existing which will load objects in an ongoing transaction and overwrite attributes with what it gets from the database; this of course assumes that your out of band changes are visible in that transaction.  

if OTOH you dont want the Session to re-query things and you'd like to inject what we refer towards as "committed" state within specific objects, the API for that is [set_committed_value](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.attributes.set_committed_value), which allows you to set the value for any attribute / collection on an object that will be represented as the current database value.


I dont know if I have more to offer at the moment since things depend highly on when and how often these OOB changes occur (within the same transaction and highly intermixed with object manipulation?   just at certain intervals) and then how tolerant the application is for the Session to refresh its data from the database, as opposed to you want to inject the new state directly into those objects.

**Comment by karolinepauls** [comment]
Thanks. I managed to get lucky with:

```
class LegacyBaseModel:

    ...

    def update_sqlalchemy(self):
        """
        Create or update the corresponding SQLAlchemy instance from `self`
        
        To be called strictly after `self` has been loaded.
        """
        from sqlalchemy.orm import make_transient_to_detached
        import the_app.models  # new SQLALchemy models

        sqla_model = the_app.models.table_to_model[self._table]

        # Create a transient instance.
        sqla_instance = sqla_model(**self._data)
        # Make it detached so it can be merged into the session withut emitting SQL.
        make_transient_to_detached(sqla_instance)
        # Merge it without reloading.
        session_sqla_instance = the_app.models.db.session.merge(sqla_instance, load=False)

        # Reference the SQLAlchemy instance so it doesn't disappear.
        self.sqla_instance = session_sqla_instance
```

Not using `set_committed_value` because that would require an instance to exist in the first place.

For now I'm trying to avoid expiring objects and emitting SQL multiple times. I'll see how far I can get.
